File: swipelibrary/src/main/java/kaufland/com/swipelibrary/dragengine/LeftDragViewEngine.java
package kaufland.com.swipelibrary.dragengine;

import android.graphics.Rect;
import android.view.View;

import kaufland.com.swipelibrary.DragView;
import kaufland.com.swipelibrary.SurfaceView;
import kaufland.com.swipelibrary.SwipeDirectionDetector;
import kaufland.com.swipelibrary.SwipeLayout;
import kaufland.com.swipelibrary.SwipeResult;
import kaufland.com.swipelibrary.SwipeState;
import kaufland.com.swipelibrary.SwipeViewLayouter;

import static kaufland.com.swipelibrary.SwipeDirectionDetector.SWIPE_DIRECTION_LEFT;
import static kaufland.com.swipelibrary.SwipeState.DragViewState.CLOSED;

/**
 * Created by sbra0902 on 29.03.17.
 */

public class LeftDragViewEngine implements DraggingEngine {

    private SurfaceView mSurfaceView;
    private DragView mDragView;

    private int mInitialXPos;

    private int mDragDistance;

    private int mIntermmediateDistance;

    public LeftDragViewEngine(DragView dragView, SurfaceView surfaceView) {
        mDragView = dragView;
        mSurfaceView = surfaceView;
    }

    @Override
    public void moveView(float offset, SurfaceView view, View changedView) {
        if (!mDragView.equals(changedView)) {
            mDragView.setX(view.getX() - mDragView.getWidth());
        }
    }

    @Override
    public void initializePosition(SwipeViewLayouter.DragDirection orientation) {

        mInitialXPos = (int) (mSurfaceView.getX() - mDragView.getWidth());
        mDragDistance = mDragView.getWidth();
        mIntermmediateDistance = mDragView.getSettlePointResourceId() != -1 ? mDragView.findViewById(mDragView.getSettlePointResourceId()).getRight() : mDragView.getWidth();

        moveToInitial();
    }

    @Override
    public void moveToInitial() {
        mDragView.setX(mInitialXPos);
    }

    @Override
    public int clampViewPositionHorizontal(View child, int left) {

        if (mDragView != null && child.equals(mDragView)) {


            if (left > 0 && !mDragView.isBouncePossible()) {
                return 0;
            }

            return left;
        }

        return 0;
    }

    @Override
    public void restoreState(SwipeState.DragViewState state, SurfaceView view) {
        switch (state) {
            case LEFT_OPEN:
                moveView(mDragDistance, view, null);
                break;
            case RIGHT_OPEN:

                moveToInitial();

                break;
            case TOP_OPEN:
                //TODO Implementation
                break;
            case BOTTOM_OPEN:
                //TODO Implementation
                break;
            default:

                moveToInitial();
                break;
        }
    }

    @Override
    public int getWidth() {
        return mDragView.getWidth();
    }

    @Override
    public int getDragDistance() {
        return mDragDistance;
    }

    @Override
    public void forceLayout() {
        mDragView.forceLayout();
    }

    @Override
    public boolean isBouncePossible() {
        return mDragView.isBouncePossible();
    }

    @Override
    public int getId() {
        return mDragView.getId();
    }

    @Override
    public boolean isDraggable() {
        return mDragView.isDraggable();
    }

    @Override
    public int getIntermmediateDistance() {
        return mIntermmediateDistance;
    }

    @Override
    public SwipeResult determineSwipeHorizontalState(float velocity, SwipeDirectionDetector swipeDirectionDetector, SwipeState swipeState, final SwipeLayout.SwipeListener swipeListener, View releasedChild) {
        if (mDragView.equals(releasedChild) && swipeDirectionDetector.getSwipeDirection() == SWIPE_DIRECTION_LEFT) {
            swipeState.setState(SwipeState.DragViewState.CLOSED);
            return new SwipeResult(-mDragView.getWidth(), new Runnable() {
                @Override
                public void run() {
                    swipeListener.onSwipeClosed(CLOSED);
                }
            });
        }

        return null;
    }

    @Override
    public DragView getDragView() {
        return mDragView;
    }
}

File: swipelibrary/src/main/java/kaufland/com/swipelibrary/dragengine/LeftDragViewEngine.kt
package kaufland.com.swipelibrary.dragengine

import android.view.View

import kaufland.com.swipelibrary.DragView
import kaufland.com.swipelibrary.SurfaceView
import kaufland.com.swipelibrary.SwipeDirectionDetector
import kaufland.com.swipelibrary.SwipeDirectionDetector.Companion.SWIPE_DIRECTION_LEFT
import kaufland.com.swipelibrary.SwipeLayout
import kaufland.com.swipelibrary.SwipeResult
import kaufland.com.swipelibrary.SwipeState
import kaufland.com.swipelibrary.SwipeViewLayouter

import kaufland.com.swipelibrary.SwipeState.DragViewState.CLOSED

/**
 * Created by sbra0902 on 29.03.17.
 */

class LeftDragViewEngine(private val mLayouter: SwipeViewLayouter) : DraggingEngine<DragView> {
    override val openOffset: Int
        get() = dragDistance

    private var mSurfaceView: SurfaceView? = null

    override var dragView: DragView? = null

    private var mInitialXPos: Int = 0

    override var dragDistance: Int = 0

    override var intermmediateDistance: Int = 0

    override fun moveView(offset: Float, view: SurfaceView, changedView: View) {
        if (dragView != changedView) {
            dragView!!.x = view.x - dragView!!.width
        }
    }

    override fun initializePosition(orientation: SwipeViewLayouter.DragDirection) {

        mSurfaceView = mLayouter.surfaceView
        dragView = mLayouter.leftDragView

        mInitialXPos = (mSurfaceView!!.x - dragView!!.width).toInt()
        dragDistance = dragView!!.width
        intermmediateDistance = if (dragView!!.settlePointResourceId != -1) dragView!!.findViewById(dragView!!.settlePointResourceId).right else dragView!!.width

        moveToInitial()
    }

    private fun moveToInitial() {
        dragView!!.x = mInitialXPos.toFloat()
    }

    override fun clampViewPositionHorizontal(child: View, left: Int): Int {

        return if (dragView != null && child == dragView) {


            if (left > 0 && !dragView!!.isBouncePossible) {
                0
            } else left

        } else 0

    }

    override fun restoreState(state: SwipeState.DragViewState, view: SurfaceView) {
        when (state) {
            SwipeState.DragViewState.LEFT_OPEN ->

                dragView!!.x = 0f
            SwipeState.DragViewState.RIGHT_OPEN -> dragView!!.x = (mLayouter.rightDragView.x - mSurfaceView!!.width.toFloat() - dragView!!.width.toFloat()).toInt().toFloat()
            SwipeState.DragViewState.TOP_OPEN -> {
            }
            SwipeState.DragViewState.BOTTOM_OPEN -> {
            }
            else -> dragView!!.x = (-dragDistance).toFloat()
        }//TODO Implementation
        //TODO Implementation
    }

    override fun determineSwipeHorizontalState(velocity: Float, swipeDirectionDetector: SwipeDirectionDetector, swipeState: SwipeState, swipeListener: SwipeLayout.SwipeListener, releasedChild: View): SwipeResult? {
        if (dragView == releasedChild && swipeDirectionDetector.swipeDirection == SWIPE_DIRECTION_LEFT && Math.abs(swipeDirectionDetector.difX) > dragDistance / 2) {
            swipeState.state = SwipeState.DragViewState.CLOSED
            return SwipeResult(-dragView!!.width, Runnable { swipeListener.onSwipeClosed(CLOSED) })
        }

        return null
    }
}
